File: Python/number_of_operations_to_make_network_connected.py
def makeConnected(n: int, connections: list) -> int:
    def find(index: int) -> int:
        if parent[index] != index:
            parent[index] = find(parent[index])
        return parent[index]
    
    def union(index1: int, index2: int):
        parent[find(index2)] = find(index1)
    
    parent = list(range(n))
    network= n
    count = 0
    for index1, index2 in connections:
        root_X = find(index1)
        root_Y = find(index2)
        if root_X != root_Y:
            union(index1, index2)
            network -= 1
        else:
            count += 1
    
    if count >= network - 1:
        return network - 1
    else:
        return -1

if __name__ == "__main__":
    print(makeConnected(n = 4, connections = [[0,1],[0,2],[1,2]]))

File: src/tmp/Java/src/com/dbc/algorithm/UnionFindSets/NumberOfOperationsToMakeNetworkConnected.java
package com.dbc.algorithm.UnionFindSets;

public class NumberOfOperationsToMakeNetworkConnected {
    public static int makeConnected(int n, int[][] connections) {
        int count = 0;
        int[] parent = new int[n];
        for (int i = 0; i < parent.length; i++) {
            parent[i] = i;
        }

        for (int[] connection : connections) {
            if (union(parent, connection[0], connection[1])) {
                n--;
            } else {
                count++;
            }
        }

        return count >= n - 1 ? n - 1 : -1;
    }

    public static int find(int[] parent, int index) {
        if (parent[index] != index) {
            parent[index] = find(parent, parent[index]);
        }
        return parent[index];
    }

    public static boolean union(int[] parent, int index1, int index2) {
        int rootX = find(parent, index1);
        int rootY = find(parent, index2);
        if (rootX != rootY) {
            parent[find(parent, index2)] = find(parent, index1);
            return true;
        } else {
            return false;
        }
    }
}
